articles/realtime.md

```diff
@@ -0,0 +1,330 @@
+The developer should avoid implementing his own spinlocks.
+The RT Preempt patch replaces much of the kernel's spinlocks with mutexes, but this might not be guaranteed on all platforms.
+
+### Avoid fork
+
+[`fork`](http://linux.die.net/man/3/memset) is not real-time safe because it is implemented using [copy-on-write](https://en.wikipedia.org/wiki/Copy-on-write).
+This means that when a forked process modifies a page of memory, it gets its own copy of that page.
+This leads to page faults!
+
+Page faults should be avoided in real-time programming, so Linux `fork`, as well as programs that call `fork`, should be avoided. 
+
+# Testing and Performance Benchmarking
+
+## cyclictest
+
+[`cyclictest`](http://manpages.ubuntu.com/manpages/trusty/man8/cyclictest.8.html) is a simple Linux command line tool for measuring the jitter of a real-time environment.
+It takes as input a number of threads, a priority for the threads, and a scheduler type.
+It spins up `n` threads that sleep regular intervals (the sleep period can also be specified from the command line). (7)
+
+For each thread, `cyclictest` measures the time between when the thread is supposed to wake up and when it actually wakes up.
+The variability of this latency is the scheduling jitter in the system.
+If there are processes with non-deterministic blocking behavior running in the system, the average latency will grow to a large number (on the order of milliseconds), since the scheduler cannot meet the deadlines of the periodically sleeping threads profiled in the program.
+
+## Instrumenting code for testing
+
+A more precise way to measure the scheduling jitter in a program is to instrument the periodic real-time update loop of existing code to record scheduling jitter.
+
+A proposed header for a minimal library for real-time code instrumentation can be found here:
+
+https://github.com/jacquelinekay/rttest/blob/master/include/rttest/rttest.h
+
+## Pagefaults
+
+The Linux system call [`getrusage`](http://linux.die.net/man/2/getrusage) returns statistics about many resource usage events relevant to real-time performance, such as minor and major pagefaults, swaps, and block I/O. It can retrieve these statistics for an entire process or for one thread.
+Thus it is simple to instrument code to check for these events. In particular, `getrusage` should be called right before and right after the real-time section of the code, and these results should be compared, since `getrusage` collects statistics about the entire duration of the thread/process.
+
+Collecting these statistics gives an indication of what events could cause the latency/scheduling jitter measured by the previously described methods.
+
+# Design Guidelines for ROS 2
+
+## Achieving real-time computation across platforms
+
+TODO: How can ROS 2 be real-time friendly and cross-platform?
+Much of the research in this document focuses on achieving real-time on POSIX-compliant Linux systems.
+
+## Implementation strategy
```

Comment:
Maybe worth mentioning explicitly. Using ROS interfaces should be real-time safe (at least within a limited context, like intra-process), setting them up needs not be. In particular, for pub-sub with dynamic-sized data, it should be possible to pre-allocate resources (e.g., resize payload arrays) at setup time.